HW2/P3/mandelbrot.pyx

```diff
@@ -41,9 +60,91 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
+    inco_real = np.real(in_coords)
+    inco_imag = np.imag(in_coords)
 
+    # 8-bit registers of 0, 2, and 4 that I use below
+    avx0 = AVX.float_to_float8(0)
+    avx1 = AVX.float_to_float8(1)
+    avx2 = AVX.float_to_float8(2)
+    avx4 = AVX.float_to_float8(4)
+    
+    with nogil:
+        for i in prange(in_coords.shape[0], schedule='static', chunksize=1, num_threads=4): 
+            # Break each row into blocks of 8
+
+            # Need number of blocks on each row:
+            blox = in_coords.shape[1]/8
+
+            # Now I parallelize within each of those blocks.  
+            for block in range(blox):              
+                
+                # Take the initial real components of c.
+                creal = AVX.make_float8(inco_real[i, 8*block + 7],
+                                        inco_real[i, 8*block + 6],
+                                        inco_real[i, 8*block + 5],
+                                        inco_real[i, 8*block + 4],
+                                        inco_real[i, 8*block + 3],
+                                        inco_real[i, 8*block + 2],
+                                        inco_real[i, 8*block + 1],
+                                        inco_real[i, 8*block + 0])
+                
+                # Take the initial imaginary components of c.  
+                cimag = AVX.make_float8(inco_imag[i, 8*block + 7],
+                                        inco_imag[i, 8*block + 6],
+                                        inco_imag[i, 8*block + 5],
+                                        inco_imag[i, 8*block + 4],
+                                        inco_imag[i, 8*block + 3],
+                                        inco_imag[i, 8*block + 2],
+                                        inco_imag[i, 8*block + 1],
+                                        inco_imag[i, 8*block + 0])
+                
+                # Starting values of zero for z and the iteration counts.
+                zreal = avx0
+                zimag = avx0              
+                itercounts = avx0
+
+                for iter in range(max_iterations):
+                
+                    # Magnitude squared (sum of squares of real and imaginary components).
+                    magsquared = AVX.mul(AVX.mul(zreal,zreal),AVX.mul(zimag,zimag))
+                     
+                    # Determine whether the magnitude squared is less than 4
+                    mask = AVX.less_than(magsquared, avx4)
+                    
+                    # If all elements are less than zero, write the iteration counts to the array.
+                    # Also, if still iterating at the last iteration, write the maximum iteration count
+                    #    to the array.  And break out of the loop.  
+                    if AVX.signs(mask)==0 or iter==(max_iterations - 1):
+                        counts_to_output(itercounts, out_counts, i, block)
+                        break
```

Comment:
Adding a "break" is important even if it will not affect the "count". I did not add break at first, so I did not see running time improvement. Good point here.